### DISCUSSION TITLE
[W600-Pico]Register programming for IO

### DISCUSSION THREAD

**Discussion by sabeehalam** [micropython/micropython#12146] [General]
Can I use registers-level-programming to poll IO in W600-Pico? If so, what register range would I need to use for the Pins?

**Comment by robert-hh** [comment]
Yes, you can. The GPIO port and other IO registers are memory mapped. The addresses can be found in the attached document, section 9.4. You can access these e.g. using the machine.mem32() method, or using viper code.
[WM_W600_Register_Manual_v1.1.zip](https://github.com/micropython/micropython/files/12239520/WM_W600_Register_Manual_v1.1.zip)

Edit: Note that the W600 port is not officially support by  MicroPython. It is supported by users of MicroPython like me.

**Reply by sabeehalam** [reply]
Thank you for the constant supporting. I just recently stumbled on the v1.20 firmware for this controller and was stuck with the old v1.10 earlier.

**Comment by sabeehalam** [comment]
Are the pin names changed for your firmware? I keep getting this error for code I was running on a dirty v1.19 firmware: AttributeError: type object 'Pin' has no attribute 'PB_09'

**Reply by robert-hh** [reply]
Yes, the methods have changed. Like the other port, the Pin module has now a cpu and board class, which directly contains the pins objects. 
```
>>> from machine import Pin
>>> dir(Pin.cpu)
['__class__', '__name__', '__bases__', '__dict__', 'PA0', 'PA1', 'PA10', 'PA11', 'PA12', 'PA13', 'PA14', 'PA15', 'PA2', 'PA3', 'PA4', 'PA5', 'PA6', 'PA7', 'PA8', 'PA9', 'PB0', 'PB1', 'PB10', 'PB11', 'PB12', 'PB13', 'PB14', 'PB15', 'PB16', 'PB17', 'PB18', 'PB19', 'PB2', 'PB20', 'PB21', 'PB22', 'PB23', 'PB24', 'PB25', 'PB26', 'PB27', 'PB28', 'PB29', 'PB3', 'PB30', 'PB31', 'PB4', 'PB5', 'PB6', 'PB7', 'PB8', 'PB9']
>>> dir(Pin.board)
['__class__', '__name__', '__bases__', '__dict__', 'LED', 'RX0', 'RX1', 'TX0', 'TX1']
>>> 
>>> type(Pin.cpu.PB9)
<class 'Pin'>
```
You can use these directly:
```
from machine import Pin
p=Pin.cpu.PB9
```
Or you can use strings or numbers for the Pin constructor. The previous Pin.PB_09 just mapped to the number 25. So in addition to use `Pin.cpu..PB9` directly, you can use:
```
p=Pin("PB9")
p=Pin(16+9)
```
The numbers 0-15 map to PA0 to PA15, the number 16-47 to PB0 to PB31.

**Reply by sabeehalam** [reply]
Ok thank you

**Comment by sabeehalam** [comment]
Is Pin.PULL_DOWN the same as Pin.PULL_NONE or is PULL_DOWN not implemented?

**Reply by robert-hh** [reply]
PULL_DOWN is not supported by the hardware, so the symbol Pin.PULL_DOWN was removed.

**Reply by sabeehalam** [reply]
OK

**Comment by sabeehalam** [comment]
Have you written any sample register level programming codes for the W600-PICO? If so can you please share them here?

**Comment by robert-hh** [comment]
> Have you written any sample register level programming codes for the W600-PICO?

No. It was not needed. What do you want to achieve? I did some for other ports. Below is a Pin toggle example in three variants. A lot of direct port access is as well in the linked file, done for STM32. https://github.com/robert-hh/SSD1963-TFT-Library-for-PyBoard-and-RP2040/blob/master/pyboard/tft_io.py The class `stm` just eases the access to addresses, mapping stm.GPIOx to the address values. 

P.S.: I re-enabled the Pin.PA_xx and Pin.PB_xx in the shared firmware images. I'll wait for dropping it until v2.x, which will have more breaking changes. 

```
from machine import Pin, mem32
from machine import Pin, mem32
gen = Pin(22, Pin.OUT)
def bop():
    while True:
        gen.on()
        gen.off()
        
def mem32bop():
    pin = (1 << 22)
    while True:
        mem32[0x3FF44008] = pin
        mem32[0x3FF4400C] = pin
        
@micropython.viper 
def viperbop():
    GPIO = ptr32(0x3FF44004)
    while True:
        GPIO[1] = 1 << 22
        GPIO[2] = 1 << 22
```
